pkg/util/image.go

```diff
@@ -3,27 +3,68 @@ package util
 import (
 	"fmt"
 
+	"github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
 	longhorntypes "github.com/longhorn/longhorn-manager/types"
+	"k8s.io/apimachinery/pkg/api/errors"
 
 	harvesterv1 "github.com/harvester/harvester/pkg/apis/harvesterhci.io/v1beta1"
+	ctllhv1 "github.com/harvester/harvester/pkg/generated/controllers/longhorn.io/v1beta2"
+
+	lhdatastore "github.com/longhorn/longhorn-manager/datastore"
+	lhutil "github.com/longhorn/longhorn-manager/util"
 )
 
-func GetBackingImageName(image *harvesterv1.VirtualMachineImage) string {
-	return fmt.Sprintf("%s-%s", image.Namespace, image.Name)
+func GetBackingImageName(backingImageCache ctllhv1.BackingImageCache, image *harvesterv1.VirtualMachineImage) (string, error) {
+	legacyName := fmt.Sprintf("%s-%s", image.Namespace, image.Name)
+	bi, err := backingImageCache.Get(LonghornSystemNamespaceName, legacyName)
+	if err == nil {
+		return bi.Name, nil
+	}
+
+	if !errors.IsNotFound(err) {
+		return "", err
+	}
+
+	return lhutil.AutoCorrectName(legacyName, lhdatastore.NameMaximumLength), nil
+}
+
+func GetBackingImageDataSourceName(backingImageCache ctllhv1.BackingImageCache, image *harvesterv1.VirtualMachineImage) (string, error) {
+	//In LH design, backingimagedatasource name is identical with backingimage
+	return GetBackingImageName(backingImageCache, image)
+}
+
+func GetBackingImage(backingImageCache ctllhv1.BackingImageCache, image *harvesterv1.VirtualMachineImage) (*v1beta2.BackingImage, error) {
+	legacyName := fmt.Sprintf("%s-%s", image.Namespace, image.Name)
+	bi, err := backingImageCache.Get(LonghornSystemNamespaceName, legacyName)
+	if err == nil {
+		return bi, nil
+	}
+
+	if !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	rectifyName := lhutil.AutoCorrectName(legacyName, lhdatastore.NameMaximumLength)
```

Comment:
We can reuse code in `GetBackingImageName`. We will call `backingImageCache.Get` again for legacy name, but it's local cache, so the extra cost can be ignored.

Reply:
How about resue `GetBackingImage() `in `GetBackingImageName()` ? Then we don't have to call backingImageCache.Get() again like this:
```
func GetBackingImageName(backingImageCache ctllhv1.BackingImageCache, image *harvesterv1.VirtualMachineImage) (string, error) {
	bi, err := GetBackingImage(backingImageCache, image)
	if err == nil {
		return bi.Name, nil
	}

	if !errors.IsNotFound(err) {
		return "", err
	}

	legacyName := fmt.Sprintf("%s-%s", image.Namespace, image.Name)
	return lhutil.AutoCorrectName(legacyName, lhdatastore.NameMaximumLength), nil
}
```

Reply:
I'm okay with it. However, do we want to create a local utility function for legacy name? We repeat it in both `GetBackingImage` and `GetBackingImageName`.